### PR TITLE
Add script fix-program-events-user

### LINKED ISSUE #34: User Program Attribute Assignment script
- Query all the events with default categoryoption for a particular program in the last 30 days.
...../api/events.json?program=fUycPoiPP58&startDate=2017-01-15&skipPaging=true (Outlet support)
....../api/events.json?program=FcSSx0KtXo9&startDate=2017-01-15&skipPaging=true (BCC)
- For each event get the username, you can get it from the field ‘storedBy’ from the dataValues object
- Then we look for the categoryoption which code is exactly the same as the storedBy value: ..../api/categoryOptions?filter=code:eq:samong
- If the category option does not exist, we iterate to the next Event. 
- If the category option exists, we edit the payload of the event modifying the attributeCategoryOptions to the id of the categoryOptions found (it contained the default categoryoption before). PUT "...../api/events/ [eventId]" Note we need to include all values on the payload. Otherwise, missing values would be imported as blank!!!!

screenshot from 2017-03-01 10 20 20

### PR DESCRIPTION
**Pull request by tokland**
Closes #34

### CODE REVIEW

File: src/scripts/fix-program-events-user/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (adrianq) [inline src/scripts/fix-program-events-user/.eslintrc.json]: @tokland Awesome! Don't know too much about this. I would like to discuss a bit about this with you on Thursday.

File: src/scripts/fix-program-events-user/config.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "api": {
+        "url": "http://localhost:8080/api/24",
```
Comment (adrianq) [inline src/scripts/fix-program-events-user/config.json]: @tokland My fault, sorry, but it was supposed to be targeting 2.25. Can you please update and test it for 2.25?

File: src/scripts/fix-program-events-user/index.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            debug("Event %s: No storedBy info found in dataValues", eventId);
+        } else if (!userId) {
+            debug("Event %s: user <%s> not found in category options", eventId, username);
+        } else if (categoryOptions == userId) {
+            debug("Event %s: user <%s> already set in category options", eventId);
+        } else {
+            const newEvent = _.defaults({attributeCategoryOptions: userId}, event);
+            const body = JSON.stringify(newEvent);
+            debug("Event %s: REPLACED <%s> by <%s> (%s)", eventId, categoryOptions, userId, username);
+            return this.request("events/" + eventId, {method: "PUT", body: body});
```
Comment (adrianq) [inline src/scripts/fix-program-events-user/index.js]: @tokland I guess what I told you about sending the whole date structure again is not needed, right? We can only send the bit that is changing, correct?

File: src/scripts/fix-program-events-user/package.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  "dependencies": {
+    "moment": "^2.15.2",
+    "request": "^2.34",
+    "request-promise": "^4.1.1",
+    "underscore": "^1.8.3",
+    "bluebird": "^3.5.0",
+    "basic-authorization-header": "^0.2.7",
+    "nconf": "^0.8.4"
+  },
+  "author": "Arnau Sanchez",
```
Comment (adrianq) [inline src/scripts/fix-program-events-user/package.json]: @tokland I know we have the name of the developers in some scripts but we have decided to set the author to 'EyeSeeTea Team' (for instance in Data Dictionary). Can you please change it accordingly?

**Review by adrianq — CHANGES_REQUESTED** [review]
@tokland Incredible work! It took me a while to understand some ES6 code as I am not familiar with it. Some minor changes requested. Can you also change the name of the script (and the description and methods) to 'user-ProgramAttribute-assignment'? Once these changes are done I will ask @josemp10 to test it and double check the dhis2 specific stuff makes sense. Thanks!

File: src/scripts/fix-program-events-user/index.js (hunk 3 above)
Comment (tokland) [inline src/scripts/fix-program-events-user/index.js]: This is sending the whole event, `_.defaults(obj1, obj2)` merges obj1 and obj2, you (or Jose) were right, if some attribute is not sent, it's removed from the DB event. That's also true for the dataValues entries, that's why its timestamp is updated.

**Comment by tokland** [comment]
I forgot, this is the debug output with `--fromDate=-100`. There are quite a lot of users nor found in the category options:

[script-output.txt](https://github.com/EyeSeeTea/organisationUnit-User-Assigment/files/826867/script-output.txt)

File: src/scripts/fix-program-events-user/index.js (hunk 3 above)
Comment (adrianq) [inline src/scripts/fix-program-events-user/index.js]: Yep, you are right. I didn't realise about the second parameter

**Review by adrianq — APPROVED** [review]

**Comment by adrianq** [comment]
@tokland Thank you very much for the changes. Everything looks good to me.

@josemp10 I have requested a review on your side about this PR. Please give it a try. If everything is working fine approve the review and I will merge the code. Otherwise, request any change. See also some notes about the PR [here](https://github.com/EyeSeeTea/organisationUnit-User-Assigment/issues/34). Note there is also a question for you in the issue. Drop me a line if you have any doubt about this. Thanks!

**Comment by josemp10** [comment]
I have tested the script in a development environment, and everything seems
to be correct. Thanks guys!

**Pull request merged**